web-common/src/features/entity-management/file-artifacts-store.ts

```diff
@@ -1,131 +1,241 @@
     });
+  }
 
-    errorsForPaths.forEach((errors, path) => {
-      createOrUpdateFileArtifact(path, (entityData: FileArtifactsData) => {
-        entityData.errors = errors;
-      });
-    });
-  },
-
-  // reducer for storing Metric artifact
-  setJSONRep(affectedPath: string, fileData: string) {
-    const jsonRepresentation = parseDocument(fileData).toJSON();
-    createOrUpdateFileArtifact(
-      affectedPath,
-      (entityData: FileArtifactsData) => {
-        entityData.jsonRepresentation = jsonRepresentation;
+  // Complex selectors based on resource API
+
+  public getResourceForFile(
+    queryClient: QueryClient,
+    instanceId: string,
+    filePath: string,
+  ) {
+    return derived(this.getResourceNameForFile(filePath), (resourceName, set) =>
+      useResource(
+        instanceId,
+        resourceName?.name as string,
+        resourceName?.kind as ResourceKind,
+        undefined,
+        queryClient,
+      ).subscribe(set),
+    ) as ReturnType<typeof useResource<V1Resource>>;
+  }
+
+  public getAllErrorsForFile(
+    queryClient: QueryClient,
+    instanceId: string,
+    filePath: string,
+  ): Readable<V1ParseError[]> {
+    return derived(
+      [
+        useProjectParser(queryClient, instanceId),
+        this.getResourceForFile(queryClient, instanceId, filePath),
+      ],
+      ([projectParser, resource]) => {
+        if (
+          projectParser.isFetching ||
+          projectParser.isError ||
+          resource.isFetching
+        ) {
+          // TODO: what should the error be for failed get resource API
+          return [];
+        }
+        return [
+          ...(
+            projectParser.data?.projectParser?.state?.parseErrors ?? []
+          ).filter(
+            (e) => e.filePath && removeLeadingSlash(e.filePath) === filePath,
+          ),
+          ...(resource.data?.meta?.reconcileError
+            ? [
+                {
+                  filePath,
+                  message: resource.data.meta.reconcileError,
+                },
+              ]
+            : []),
+        ];
       },
+      [],
     );
-  },
+  }
 
-  setIsReconciling(affectedPath: string, isReconciling: boolean) {
-    createOrUpdateFileArtifact(
-      affectedPath,
-      (entityData: FileArtifactsData) => {
-        entityData.isReconciling = isReconciling;
-      },
+  public getFileHasErrors(
+    queryClient: QueryClient,
+    instanceId: string,
+    filePath: string,
+  ) {
+    return derived(
+      this.getAllErrorsForFile(queryClient, instanceId, filePath),
+      (errors) => errors.length > 0,
     );
-  },
-};
-
-export type FileArtifactsStore = Readable<FileArtifactsState> &
-  typeof fileArtifactsEntitiesReducers;
-
-export const fileArtifactsStore: FileArtifactsStore = {
-  subscribe,
-  ...fileArtifactsEntitiesReducers,
-};
-
-export function getIsFileReconcilingStore(file: string) {
-  return derived(
-    fileArtifactsStore,
-    ($store) => $store.entities[file]?.isReconciling,
-  );
+  }
```

Comment:
I'd also expect all these methods to be on the `FileArtifact` class itself, not on the parent container